golang/src/user-preferences/main.go

```diff
@@ -0,0 +1,393 @@
+package main
+
+import (
+	"configurate"
+	"database/sql"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"logcabin"
+	"net/http"
+	"os"
+	"queries"
+	"strings"
+
+	"github.com/gorilla/mux"
+	_ "github.com/lib/pq"
+)
+
+var (
+	version = flag.Bool("version", false, "Print the version information")
+	cfgPath = flag.String("config", "/etc/iplant/de/jobservices.yml", "The path to the config file")
+	port    = flag.String("port", "60000", "The port number to listen on")
+	gitref  string
+	appver  string
+	builtby string
+)
+
+func init() {
+	flag.Parse()
+}
+
+// AppVersion prints the version information to stdout
+func AppVersion() {
+	if appver != "" {
+		fmt.Printf("App-Version: %s\n", appver)
+	}
+	if gitref != "" {
+		fmt.Printf("Git-Ref: %s\n", gitref)
+	}
+	if builtby != "" {
+		fmt.Printf("Built-By: %s\n", builtby)
+	}
+}
+
+// App defines the interface for a user-preferences application
+type App interface {
+	Greeting(http.ResponseWriter, *http.Request)
+	GetRequest(http.ResponseWriter, *http.Request)
+	PutRequest(http.ResponseWriter, *http.Request)
+	PostRequest(http.ResponseWriter, *http.Request)
+	DeleteRequest(http.ResponseWriter, *http.Request)
+}
+
+// UserPreferencesRecord represents a user preferences stored in the database
+type UserPreferencesRecord struct {
+	ID          string
+	Preferences string
+	UserID      string
+}
+
+// convert makes sure that the JSON has the correct format. "wrap" tells convert
+// whether to wrap the object in a map with "session" as the key.
```

Comment:
this is incorrect documentation, it's not "session" here. might be nice to move a generalized version of this into a shared namespace, I bet the other two services will also want this and no sense providing more ways to produce copy-paste errors (there's also a reference to 'session' in a few comments below)